thonny/base_file_browser.py

```diff
@@ -863,12 +880,132 @@ def should_open_name_in_thonny(self, name):
         ext = self.get_extension_from_name(name)
         return get_workbench().get_option(get_file_handler_conf_key(ext), "system") == "thonny"
 
+    def supports_new_file(self):
+        return False
+
+    def create_new_file(self):
+        raise NotImplementedError()
+
+    def get_selected_file(self):
+        selection = self.get_selection_info(True)
+        if not selection or len(selection["paths"]) > 1:
+            return
+        return selection["paths"][0]
+
+    def supports_rename(self):
+        return False
+
+    def rename_file(self):
+        old_name = self.get_selected_file()
+        if old_name is None:
+            return
+        file_name = os.path.basename(old_name)
+
+        new_name = simpledialog.askstring(
+            tr("Rename '%s'") % file_name, tr("Enter new name"), initialvalue=file_name, parent=self
+        )
+        if new_name:
+            rc = self.perform_rename(old_name, new_name)
+            if rc:
+                messagebox.showerror(tr("Rename of '%s' failed") % old_name, rc, parent=self)
+
+    def perform_rename(self, old_name, new_name):
+        raise Exception("overload this in subclass")
+
+    def supports_copypaste(self):
+        return self.copypaste
+
+    def cut_files(self):
+        assert self.copypaste
+        self.copypaste.cut()
+
+    def copy_files(self):
+        assert self.copypaste
+        self.copypaste.copy()
+
+    def paste_files(self):
+        assert self.copypaste
+        target = self.get_selected_file()
+        assert target
+        if os.path.isfile(target):
+            target = os.path.dirname(target)
+        self.copypaste.paste(target)
+
+
+class CopyPaste(object):
+    def __init__(self, filebrowser):
+        self.fb = filebrowser
+        self.tree = filebrowser.tree
+        self.reset()
+
+    def reset(self):
+        self.mode = None
+        self.paths = []
+
+    def get_selection_paths(self):
+        selection = self.fb.get_selection_info(True)
+        if selection:
+            return selection["paths"]
+        return []
+
+    def has_selection(self):
+        return len(self.paths) > 0
+
+    def dirname(self, path):
+        if os.path.isfile(path):
+            path = os.path.dirname(path)
+        return path
+
+    def conflicts(self, target):
+        target = self.dirname(target)
+        folder = set(map(lambda x: self.dirname(x), self.paths))
```

Comment:
Do I understand properly, that you want to avoid the situation where any of the source files gets pasted to where it already is? 

I think this is valid, but you should also forbid cut-pasting into a directory which is _below_ any of the selected items.